File: src/commands/doctor.ts
import { spawnSync } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import chalk from 'chalk';

import { printJsonCommand } from '../cli/json.js';
import { type ConfigLoad, readConfig, resolveMachineId } from '../config.js';
import { findDuplicateMachineDays } from '../data/duplicateMachines.js';
import { isRecord } from '../data/guards.js';
import { INIT_HINT } from '../data/messages.js';
import type { MachineFile } from '../data/types.js';
import { pad } from '../display/format.js';
import { errorMessage } from '../errors.js';
import { isCloned, listDataFiles, LOCAL_REPO, readDataFile } from '../git.js';
import { log } from '../output.js';
import { CLAUDE_PRICING_BY_ID } from '../pricing/claude.js';
import { CODEX_PRICING_BY_ID } from '../pricing/codex.js';
import { getClaudePaths } from '../readers/claude.js';
import { getCodexPaths } from '../readers/codex.js';
import { getCursorStateDatabasePath, readCursorAuthState } from '../readers/cursor/auth.js';
import { jsonlSourceSummary } from '../readers/paths.js';

interface DoctorOptions {
  pricingCheck?: boolean;
  json?: boolean;
}

type CheckStatus = 'ok' | 'warn' | 'fail';

interface CheckResult {
  status: CheckStatus;
  label: string;
  detail: string;
}

/**
 * Status text and its color, kept apart so the column can be padded on the
 * plain text.
 *
 * Padding the colored string instead measured the ANSI escapes as content, so
 * `padEnd` was a no-op on a TTY and the columns only lined up when color was
 * disabled. `display/terminalTable.ts` already pads before styling.
 */
const STATUS_STYLE: Record<CheckStatus, { text: string; color: (value: string) => string }> = {
  ok: { text: 'OK', color: (value) => chalk.green(value) },
  warn: { text: 'WARN', color: (value) => chalk.yellow(value) },
  fail: { text: 'FAIL', color: (value) => chalk.red(value) },
};

/** Width of the widest status text, so every row's label starts in one column. */
const STATUS_COLUMN_WIDTH = 4;

function statusLabel(status: CheckStatus): string {
  const style = STATUS_STYLE[status];
  return style.color(pad(style.text, STATUS_COLUMN_WIDTH, 'left'));
}

function parseMajor(version: string): number {
  return Number(version.split('.', 1)[0] ?? 0);
}

interface CommandRunResult {
  ok: boolean;
  output: string;
}

function runCommand(command: string, arguments_: string[], cwd?: string): CommandRunResult {
  const result = spawnSync(command, arguments_, {
    cwd,
    encoding: 'utf8',
    stdio: 'pipe',
    timeout: 10_000,
  });
  const output = [result.stderr, result.stdout]
    .filter((chunk): chunk is string => typeof chunk === 'string' && chunk.trim().length > 0)
    .join('\n')
    .trim();
  return { ok: result.status === 0, output };
}

function commandCheck(
  label: string,
  command: string,
  arguments_: string[],
  options: {
    cwd?: string;
    okStatus?: CheckStatus;
    failStatus?: CheckStatus;
    okDetail: string | ((output: string) => string);
    failDetail: string | ((output: string) => string);
  },
): CheckResult {
  const run = runCommand(command, arguments_, options.cwd);
  const detail = run.ok
    ? typeof options.okDetail === 'function'
      ? options.okDetail(run.output)
      : options.okDetail
    : typeof options.failDetail === 'function'
      ? options.failDetail(run.output)
      : options.failDetail;
  return {
    status: run.ok ? (options.okStatus ?? 'ok') : (options.failStatus ?? 'fail'),
    label,
    detail,
  };
}

async function sourceCheck(label: string, roots: string[]): Promise<CheckResult> {
  const { existing, fileCount } = await jsonlSourceSummary(roots);
  if (fileCount > 0) {
    return {
      status: 'ok',
      label,
      detail: `${String(fileCount)} JSONL file(s) across ${String(existing.length)} existing path(s)`,
    };
  }
  if (existing.length > 0) {
    return {
      status: 'warn',
      label,
      detail: `paths exist but no JSONL files were found: ${existing.join(', ')}`,
    };
  }
  return {
    status: 'warn',
    label,
    detail: `no source paths found; checked ${roots.join(', ')}`,
  };
}

/** Whether `directory` is the aitrack repo, which is what carries the script. */
function isAitrackCheckout(directory: string): boolean {
  try {
    const parsed: unknown = JSON.parse(readFileSync(join(directory, 'package.json'), 'utf8'));
    if (!isRecord(parsed)) return false;
    return (
      parsed.name === 'aitrack' && isRecord(parsed.scripts) && 'pricing:check' in parsed.scripts
    );
  } catch {
    return false;
  }
}

function pricingCheck(options: DoctorOptions): CheckResult {
  if (!options.pricingCheck) {
    return {
      status: 'ok',
      label: 'Pricing tables',
      detail: `${String(Object.keys(CLAUDE_PRICING_BY_ID).length)} Claude and ${String(
        Object.keys(CODEX_PRICING_BY_ID).length,
      )} Codex model entries bundled; run doctor --pricing-check for drift check`,
    };
  }

  // The drift check shells out to `pnpm run pricing:check`, which only exists
  // in a source checkout. A published install has neither the script nor
  // necessarily pnpm, and any unrelated project's package.json would previously
  // get this far and fail with a confusing pnpm error.
  if (!isAitrackCheckout(process.cwd())) {
    return {
      status: 'warn',
      label: 'Pricing drift',
      detail:
        'not an aitrack source checkout; run doctor --pricing-check from the aitrack repo root',
    };
  }

  const run = runCommand('pnpm', ['run', 'pricing:check'], process.cwd());
  return run.ok
    ? { status: 'ok', label: 'Pricing drift', detail: 'pnpm run pricing:check passed' }
    : {
        status: 'warn',
        label: 'Pricing drift',
        detail: run.output
          ? `pnpm run pricing:check failed: ${run.output}`
          : 'pnpm run pricing:check did not pass; run from the aitrack repo root and inspect its output',
      };
}

async function cursorCheck(): Promise<CheckResult> {
  const stateDb = getCursorStateDatabasePath();
  if (!stateDb) {
    return {
      status: 'warn',
      label: 'Cursor source',
      detail: 'state.vscdb not found; Cursor usage will be skipped unless configured',
    };
  }

  try {
    const auth = await readCursorAuthState(stateDb);
    return auth.accessToken
      ? { status: 'ok', label: 'Cursor source', detail: `auth token found in ${stateDb}` }
      : {
          status: 'warn',
          label: 'Cursor source',
          detail: `state DB found but no access token was present: ${stateDb}`,
        };
  } catch (error) {
    const message = errorMessage(error);
    return { status: 'warn', label: 'Cursor source', detail: message };
  }
}

// One physical machine synced under two identities (e.g. after a hostname or
// machineId change that left the old data file behind) gets counted twice in
// every aggregate. The tell is the same day carrying a byte-identical payload
// under more than one machine file — real distinct machines never collide.
export function duplicateMachineCheck(): CheckResult {
  const machines = listDataFiles()
    .map((filePath) => readDataFile(filePath))
    .filter((machine): machine is MachineFile => machine !== null);

  const duplicates = findDuplicateMachineDays(machines);
  if (duplicates.days.length === 0) {
    return {
      status: 'ok',
      label: 'Machine identities',
      detail: `${String(machines.length)} machine(s), no duplicated days`,
    };
  }

  return {
    status: 'warn',
    label: 'Machine identities',
    detail:
      `${String(duplicates.days.length)} day(s) are recorded identically under multiple machines (${duplicates.machines.join(', ')}) — ` +
      'totals are inflated. These are likely one machine synced under several ids; ' +
      'merge them into one data file.',
  };
}

/**
 * Report the config, distinguishing a file that is absent from one that is
 * present but broken — which used to look identical here, so a corrupt config
 * was diagnosed as "no config found" and the advice was to re-run init.
 */
function configCheck(loaded: ConfigLoad): CheckResult {
  if (loaded.status === 'ok') {
    return {
      status: 'ok',
      label: 'Config',
      detail: `repoUrl=${loaded.config.repoUrl || '(empty)'}, machineId=${resolveMachineId(loaded.config)}`,
    };
  }
  if (loaded.status === 'invalid') {
    return {
      status: 'fail',
      label: 'Config',
      detail: `config file is ${loaded.reason}; fix it or re-run ${INIT_HINT}`,
    };
  }
  return {
    status: 'warn',
    label: 'Config',
    detail: `no config found; run ${INIT_HINT} for sync`,
  };
}

async function collectChecks(options: DoctorOptions): Promise<CheckResult[]> {
  const loadedConfig = readConfig();
  const checks: CheckResult[] = [];

  checks.push({
    status: parseMajor(process.versions.node) >= 24 ? 'ok' : 'fail',
    label: 'Node.js',
    detail: `${process.version} (requires >=24)`,
  });
  checks.push(
    commandCheck('git', 'git', ['--version'], {
      okDetail: 'available on PATH',
      failDetail: 'not available on PATH',
    }),
  );
  checks.push(configCheck(loadedConfig));
  checks.push({
    status: isCloned() ? 'ok' : 'warn',
    label: 'Local repo',
    detail: isCloned() ? LOCAL_REPO : 'not cloned; local preview still works',
  });

  if (isCloned()) {
    checks.push(duplicateMachineCheck());
    checks.push(
      commandCheck('Repo health', 'git', ['status', '--short'], {
        cwd: LOCAL_REPO,
        okDetail: 'git status succeeded',
        failDetail: (output) =>
          output ? `git status failed: ${output}` : 'git status failed in local repo',
      }),
    );
    checks.push(
      commandCheck('Remote push', 'git', ['push', '--dry-run'], {
        cwd: LOCAL_REPO,
        okStatus: 'ok',
        failStatus: 'warn',
        okDetail: 'git push --dry-run succeeded',
        failDetail: (output) =>
          output
            ? `git push --dry-run failed: ${output}`
            : 'git push --dry-run failed; check remote access and branch tracking',
      }),
    );
  }

  // Three independent probes that each hit the filesystem or Cursor's database.
  checks.push(
    ...(await Promise.all([
      sourceCheck('Claude Code source', getClaudePaths()),
      sourceCheck('Codex source', getCodexPaths()),
      cursorCheck(),
    ])),
  );
  checks.push(pricingCheck(options));

  return checks;
}

export async function doctorCommand(options: DoctorOptions = {}): Promise<void> {
  const checks = await collectChecks(options);

  if (options.json) {
    printJsonCommand('doctor', {
      checks: checks.map((check) => ({
        status: check.status,
        label: check.label,
        detail: check.detail,
      })),
      hasFailures: checks.some((check) => check.status === 'fail'),
    });
  } else {
    log.info(chalk.bold('aitrack doctor'));
    for (const check of checks) {
      log.info(`${statusLabel(check.status)}  ${check.label}: ${check.detail}`);
    }
  }

  if (checks.some((check) => check.status === 'fail')) {
    process.exitCode = 1;
  }
}

File: crates/aitrack-core/src/doctor.rs
//! Doctor command implementation - diagnostic reports.

use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::process::{Command, Stdio};

use crate::config::{ConfigLoad, read_config, resolve_machine_id};
use crate::data::duplicate_machines::find_duplicate_machine_days;
use crate::data::types::MachineFile;
use crate::git::is_cloned;
use crate::paths::local_repo;
use crate::store::list_data_files;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum CheckStatus {
    Ok,
    Warn,
    Fail,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckResult {
    pub status: CheckStatus,
    pub label: String,
    pub detail: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DoctorReport {
    pub checks: Vec<CheckResult>,
    pub has_failures: bool,
}

fn run_command(command: &str, args: &[&str], cwd: Option<&std::path::Path>) -> (bool, String) {
    let mut cmd = Command::new(command);
    cmd.args(args).stdout(Stdio::piped()).stderr(Stdio::piped());

    if let Some(dir) = cwd {
        cmd.current_dir(dir);
    }

    match cmd.output() {
        Ok(output) => {
            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
            let combined = if !stderr.is_empty() && !stdout.is_empty() {
                format!("{}\n{}", stderr, stdout)
            } else if !stderr.is_empty() {
                stderr
            } else {
                stdout
            };

            (output.status.success(), combined)
        }
        Err(e) => (false, e.to_string()),
    }
}

fn config_check() -> CheckResult {
    match read_config() {
        ConfigLoad::Ok { config } => CheckResult {
            status: CheckStatus::Ok,
            label: "Config".to_string(),
            detail: format!(
                "repoUrl={}, machineId={}",
                config.repo_url,
                resolve_machine_id(&config).unwrap_or_else(|_| "unknown".to_string())
            ),
        },
        ConfigLoad::Invalid { reason } => CheckResult {
            status: CheckStatus::Fail,
            label: "Config".to_string(),
            detail: format!(
                "config file is {}; fix it or re-run: npx aitrack init",
                reason
            ),
        },
        ConfigLoad::Missing => CheckResult {
            status: CheckStatus::Warn,
            label: "Config".to_string(),
            detail: "no config found; run npx aitrack init for sync".to_string(),
        },
    }
}

fn duplicate_machine_check() -> CheckResult {
    let machines: Vec<MachineFile> = list_data_files()
        .unwrap_or_default()
        .iter()
        .filter_map(|path| crate::store::read_data_file(path).ok().flatten())
        .collect();

    if machines.is_empty() {
        return CheckResult {
            status: CheckStatus::Ok,
            label: "Machine identities".to_string(),
            detail: "0 machine(s), no duplicated days".to_string(),
        };
    }

    let duplicates = find_duplicate_machine_days(&machines);

    if duplicates.days.is_empty() {
        return CheckResult {
            status: CheckStatus::Ok,
            label: "Machine identities".to_string(),
            detail: format!("{} machine(s), no duplicated days", machines.len()),
        };
    }

    CheckResult {
        status: CheckStatus::Warn,
        label: "Machine identities".to_string(),
        detail: format!(
            "{} day(s) are recorded identically under multiple machines ({}) — \
             totals are inflated. These are likely one machine synced under several ids; \
             merge them into one data file.",
            duplicates.days.len(),
            duplicates.machines.join(", ")
        ),
    }
}

fn node_version_check() -> CheckResult {
    let (ok, output) = run_command("node", &["--version"], None);

    if !ok {
        return CheckResult {
            status: CheckStatus::Fail,
            label: "Node.js".to_string(),
            detail: "not available on PATH".to_string(),
        };
    }

    // Parse version (e.g., "v24.0.0" -> 24)
    let major = output
        .trim()
        .trim_start_matches('v')
        .split('.')
        .next()
        .and_then(|s| s.parse::<u32>().ok())
        .unwrap_or(0);

    CheckResult {
        status: if major >= 24 {
            CheckStatus::Ok
        } else {
            CheckStatus::Fail
        },
        label: "Node.js".to_string(),
        detail: format!("{} (requires >=24)", output.trim()),
    }
}

fn git_check() -> CheckResult {
    let (ok, _) = run_command("git", &["--version"], None);

    CheckResult {
        status: if ok {
            CheckStatus::Ok
        } else {
            CheckStatus::Fail
        },
        label: "git".to_string(),
        detail: if ok {
            "available on PATH".to_string()
        } else {
            "not available on PATH".to_string()
        },
    }
}

fn local_repo_check() -> CheckResult {
    let cloned = is_cloned();

    CheckResult {
        status: if cloned {
            CheckStatus::Ok
        } else {
            CheckStatus::Warn
        },
        label: "Local repo".to_string(),
        detail: if cloned {
            local_repo().display().to_string()
        } else {
            "not cloned; local preview still works".to_string()
        },
    }
}

fn repo_health_check() -> Option<CheckResult> {
    if !is_cloned() {
        return None;
    }

    let (ok, output) = run_command("git", &["status", "--short"], Some(&local_repo()));

    Some(CheckResult {
        status: if ok {
            CheckStatus::Ok
        } else {
            CheckStatus::Fail
        },
        label: "Repo health".to_string(),
        detail: if ok {
            "git status succeeded".to_string()
        } else if !output.is_empty() {
            format!("git status failed: {}", output)
        } else {
            "git status failed in local repo".to_string()
        },
    })
}

fn remote_push_check() -> Option<CheckResult> {
    if !is_cloned() {
        return None;
    }

    let (ok, output) = run_command("git", &["push", "--dry-run"], Some(&local_repo()));

    Some(CheckResult {
        status: if ok {
            CheckStatus::Ok
        } else {
            CheckStatus::Warn
        },
        label: "Remote push".to_string(),
        detail: if ok {
            "git push --dry-run succeeded".to_string()
        } else if !output.is_empty() {
            format!("git push --dry-run failed: {}", output)
        } else {
            "git push --dry-run failed; check remote access and branch tracking".to_string()
        },
    })
}

/// Generate a full doctor report with all diagnostic checks.
pub fn doctor_report() -> DoctorReport {
    let mut checks = vec![
        node_version_check(),
        git_check(),
        config_check(),
        local_repo_check(),
    ];

    if is_cloned() {
        checks.push(duplicate_machine_check());
        if let Some(check) = repo_health_check() {
            checks.push(check);
        }
        if let Some(check) = remote_push_check() {
            checks.push(check);
        }
    }

    let has_failures = checks.iter().any(|c| c.status == CheckStatus::Fail);

    DoctorReport {
        checks,
        has_failures,
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MachineSummary {
    pub hostname: String,
    pub last_updated: String,
    pub days: usize,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub total_tokens: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost_usd: Option<f64>,
    pub providers: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_day: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_day: Option<String>,
}

fn summarize_machine(file: &MachineFile) -> MachineSummary {
    let mut input_tokens = 0u64;
    let mut output_tokens = 0u64;
    let mut cost_usd = 0.0;
    let mut has_cost = false;
    let mut providers = HashSet::new();
    let mut day_keys = Vec::new();

    for (date, day_providers) in &file.days {
        let mut day_has_tokens = false;

        for (provider_key, provider_data) in day_providers {
            providers.insert(provider_key.clone());
            input_tokens += provider_data.totals.input_tokens;
            output_tokens += provider_data.totals.output_tokens;

            if let Some(cost) = provider_data.totals.cost_usd {
                cost_usd += cost;
                has_cost = true;
            }

            if provider_data.totals.input_tokens + provider_data.totals.output_tokens > 0 {
                day_has_tokens = true;
            }
        }

        if day_has_tokens {
            day_keys.push(date.clone());
        }
    }

    day_keys.sort();

    let mut provider_list: Vec<String> = providers.into_iter().collect();
    provider_list.sort();

    MachineSummary {
        hostname: file.hostname.clone(),
        last_updated: file.last_updated.clone(),
        days: day_keys.len(),
        input_tokens,
        output_tokens,
        total_tokens: input_tokens + output_tokens,
        cost_usd: if has_cost { Some(cost_usd) } else { None },
        providers: provider_list,
        first_day: day_keys.first().cloned(),
        last_day: day_keys.last().cloned(),
    }
}

/// Generate a machines report with summaries sorted by total tokens.
pub fn machines_report() -> Vec<MachineSummary> {
    let machines: Vec<MachineFile> = list_data_files()
        .unwrap_or_default()
        .iter()
        .filter_map(|path| crate::store::read_data_file(path).ok().flatten())
        .collect();

    let mut summaries: Vec<MachineSummary> = machines.iter().map(summarize_machine).collect();

    summaries.sort_by_key(|a| std::cmp::Reverse(a.total_tokens));

    summaries
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_doctor_report_runs() {
        let report = doctor_report();
        assert!(!report.checks.is_empty());
    }

    #[test]
    fn test_machines_report_empty() {
        let report = machines_report();
        // Just ensure it doesn't crash
        assert!(report.is_empty() || !report.is_empty());
    }
}
